### PR TITLE
Add number theory and homomorphic encryption library packages. helib, ntl, palisade-development, seal.

### PR DESCRIPTION
**Pull request by wohlbier**
Add number theory and homomorphic encryption library packages. helib, ntl, palisade-development-git, seal.

### CODE REVIEW

**Comment by wohlbier** [comment]
@tldahlgren I opened a new PR for these packages.

File: var/spack/repos/builtin/packages/palisade-development-git/package.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PalisadeDevelopmentGit(CMakePackage):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/palisade-development-git/package.py]: Usually we don't append "git" to packages that can be downloaded from git. Sometimes packages have a mixture of archive versions and SCM versions, which can both be managed in the same package. Therefore I think this should be called `PalisadeDevelopment` (`Palisade` would also work - I don't think that's "taken").

File: var/spack/repos/builtin/packages/ntl/package.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class Ntl(Package):
+    """
+    NTL  -- a library for doing numbery theory
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/ntl/package.py]: Typo: "numbery"

File: var/spack/repos/builtin/packages/ntl/package.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    def configure(self, spec, prefix):
+        with working_dir('src'):
+            configure = Executable('./configure')
+            configure(*self.configure_args())
+
+    def build(self, spec, prefix):
+        with working_dir('src'):
+            make()
+
+    def install(self, spec, prefix):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/ntl/package.py:60]: I think you wouldn't have to implement `configure`/`build`/`install` if you inherited `AutotoolsPackage`, would you be able to try that?

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a couple minor requests but other than that this looks good.

File: var/spack/repos/builtin/packages/ntl/package.py (hunk 3 above)
Comment (wohlbier) [reply var/spack/repos/builtin/packages/ntl/package.py:60]: The package has a nonstandard directory layout. If I inherit from `AutotoolsPackage` is there a way to use `src` as the working directory?

Comment (wohlbier) [reply var/spack/repos/builtin/packages/ntl/package.py:60]: @scheibelp I pushed the first two changes. I will experiment with inheriting from `AutotoolsPackage`, but I don't know yet if I'll be able to set `working_dir('src')`.

Comment (wohlbier) [reply var/spack/repos/builtin/packages/ntl/package.py:60]: @scheibelp It turns out that ntl isn't actually an autotools package. To see this look at the configure script https://github.com/libntl/ntl/blob/main/src/configure
In this case, the documentation suggests using `Package`.  See the "Warning": https://spack.readthedocs.io/en/latest/build_systems/autotoolspackage.html

File: var/spack/repos/builtin/packages/ntl/package.py (hunk 2 above)
Comment (wohlbier) [reply var/spack/repos/builtin/packages/ntl/package.py]: Done

File: var/spack/repos/builtin/packages/palisade-development-git/package.py (hunk 1 above)
Comment (wohlbier) [reply var/spack/repos/builtin/packages/palisade-development-git/package.py]: Done

File: var/spack/repos/builtin/packages/ntl/package.py (hunk 3 above)
Comment (scheibelp) [reply var/spack/repos/builtin/packages/ntl/package.py:60]: Thanks for checking that! This should probably stay `Package` then.

**Review by scheibelp — APPROVED** [review]

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**